### PR TITLE
[mono][wasm] Allow AOTing methods with catch/filter clauses.

### PR DESCRIPTION
**Pull request by vargaz**
This works as follows:
* During EH, when a catch clause is found in an AOTed method, the
  EH state is saved into TLS and a c++ exception is thrown.
* The C++ exception is caught by the AOTed method, and the landing
  pad code calls mono_llvm_resume_exception_il_state ().
* That call will run the catch clause and the rest of the method
  code using the interpreter, storing the possible return value
  back into the AOTed method's stack frame.
* After the call, the method skips the rest of its code, and
  returns immediately to its caller.

### CODE REVIEW

**Comment by vargaz** [comment]
Running the console benchmark:

BEFORE:
```
|        Exceptions, NoExceptionHandling |     0.4793us |
|                   Exceptions, TryCatch |     0.4888us |
|              Exceptions, TryCatchThrow |     0.0146ms |
|             Exceptions, TryCatchFilter |     0.4176us |
|       Exceptions, TryCatchFilterInline |     0.2308us |
|        Exceptions, TryCatchFilterThrow |     0.0169ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0170ms |
|         Json, non-ASCII text serialize |     0.8488ms |
|       Json, non-ASCII text deserialize |     1.8116ms |
|                  Json, small serialize |     0.0936ms |
|                Json, small deserialize |     0.1772ms |
|                  Json, large serialize |    16.0582ms |
|                Json, large deserialize |    50.0947ms |
```
AFTER:
```
|        Exceptions, NoExceptionHandling |     0.2430us |
|                   Exceptions, TryCatch |     0.3731us |
|              Exceptions, TryCatchThrow |     0.0468ms |
|             Exceptions, TryCatchFilter |     0.5380us |
|       Exceptions, TryCatchFilterInline |     0.3169us |
|        Exceptions, TryCatchFilterThrow |     0.0185ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0177ms |
|         Json, non-ASCII text serialize |     0.9475ms |
|       Json, non-ASCII text deserialize |     1.7774ms |
|                  Json, small serialize |     0.0790ms |
|                Json, small deserialize |     0.1241ms |
|                  Json, large serialize |    16.3476ms |
|                Json, large deserialize |    34.5957ms |
```

**Comment by vargaz** [comment]
The wasm failures are caused by emscripten OOM.

**Comment by vargaz** [comment]
With filter clauses:
```
| measurement | time |
|-:|-:|
|        Exceptions, NoExceptionHandling |     0.2323us |
|                   Exceptions, TryCatch |     0.3679us |
|              Exceptions, TryCatchThrow |     0.0163ms |
|             Exceptions, TryCatchFilter |     0.3177us |
|       Exceptions, TryCatchFilterInline |     0.2124us |
|        Exceptions, TryCatchFilterThrow |     0.0237ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0201ms |
|         Json, non-ASCII text serialize |     0.4899ms |
|       Json, non-ASCII text deserialize |     1.7639ms |
|                  Json, small serialize |     0.0513ms |
|                Json, small deserialize |     0.0795ms |
|                  Json, large serialize |    14.6339ms |
|                Json, large deserialize |    22.0000ms |
```

**Comment by lewing** [comment]
@lambdageek @BrzVlad this looks ready for review.

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -7433,32 +7478,38 @@ interp_run_finally_with_il_state (gpointer il_state_ptr, int clause_index, gpoin
...
 		if (il_state->data [locals_start + i])
 			stackval_from_data (header->locals [i], (stackval*)(frame_locals (&frame) + imethod->local_offsets [i]), il_state->data [locals_start + i], FALSE);
 	}
 
 	memset (&clause_args, 0, sizeof (FrameClauseArgs));
 	clause_args.start_with_ip = (const guint16*)handler_ip;
-	clause_args.end_at_ip = (const guint16*)handler_ip_end;
-	clause_args.exit_clause = clause_index;
+	if (clause_type == MONO_EXCEPTION_CLAUSE_NONE || clause_type == MONO_EXCEPTION_CLAUSE_FILTER)
+		clause_args.end_at_ip = (const guint16*)clause_args.start_with_ip + 0xffffff;
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/interp.c:7490]: Is this saying if we're in a filter or a catch clause to run until the end of the method?

Comment (vargaz) [reply src/mono/mono/mini/interp/interp.c:7490]: Yes.

Comment (vargaz) [reply src/mono/mono/mini/interp/interp.c:7490]: For filter clauses, it will run until ENDFILTER.

**Review by lambdageek — COMMENTED** [review]
LGTM; but I'd like @BrzVlad 's opinion

File: src/mono/mono/mini/interp/interp.c [hunk 2]
```diff
@@ -3897,6 +3926,8 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 				frame->state.ip = ip;
 				error_init_reuse (error);
 				do_jit_call (context, (stackval*)(locals + return_offset), (stackval*)(locals + call_args_offset), frame, cmethod, error);
+				if (context->pending_unwind)
+					return;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:3930]: We shouldn't simply return from `interp_exec_method`. Interpreter frames should be exited one by one. By simply returning we fail to restore the `context->data_stack`. `pending_unwind` should follow the same behavior as `has_resume_state` when it comes to exiting interpreter frames, they should probably be partly unified.

**Comment by vargaz** [comment]
CI failures look unrelated.

File: src/mono/mono/mini/interp/interp.c [hunk 3]
```diff
@@ -2641,9 +2651,21 @@ do_jit_call (ThreadContext *context, stackval *ret_sp, stackval *sp, InterpFrame
...
 			return;
-		MonoObject *obj = mono_llvm_load_exception ();
+		MonoJitTlsData *jit_tls = mono_get_jit_tls ();
+		if (jit_tls->resume_state.il_state) {
+			/*
+			 * This c++ exception is going to be caught by an AOTed frame above us.
+			 * We can't rethrow here, since that will skip the cleanup of the
+			 * interpreter stack space etc. So instruct the interpreter to unwind.
+			 */
+			printf ("AAA\n");
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:2661]: left-over printf

**Review by BrzVlad — APPROVED** [review]

**Comment by vargaz** [comment]
CI failures look unrelated.

**Pull request merged**